pkg/discovery/discovery.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap.com/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/client/clientset/versioned"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// TiDBDiscovery service helps new PD member to discover all other members in cluster bootstrap phase.
+type TiDBDiscovery struct {
+	cli       versioned.Interface
+	lock      sync.Mutex
+	clusters  map[string]*clusterInfo
+	tcGetFn   func(ns, tcName string) (*v1alpha1.TidbCluster, error)
+	pdControl controller.PDControlInterface
+}
+
+type clusterInfo struct {
+	resourceVersion string
+	peers           map[string]struct{}
+}
+
+// NewTiDBDiscovery returns a TiDBDiscovery
+func NewTiDBDiscovery(cli versioned.Interface) *TiDBDiscovery {
+	td := &TiDBDiscovery{
+		cli:       cli,
+		pdControl: controller.NewDefaultPDControl(),
+		clusters:  map[string]*clusterInfo{},
+	}
+	td.tcGetFn = td.realTCGetFn
+	return td
+}
+
+func (td *TiDBDiscovery) Discover(advertisePeerUrl string) (string, error) {
+	td.lock.Lock()
+	defer td.lock.Unlock()
+
+	if advertisePeerUrl == "" {
+		return "", fmt.Errorf("advertisePeerUrl is empty")
+	}
+	strArr := strings.Split(advertisePeerUrl, ".")
+	if len(strArr) != 4 {
+		return "", fmt.Errorf("advertisePeerUrl format is wrong: %s", advertisePeerUrl)
+	}
+
+	podName, peerServiceName, ns := strArr[0], strArr[1], strArr[2]
+	tcName := strings.TrimSuffix(peerServiceName, "-pd-peer")
+	podNamespace := os.Getenv("MY_POD_NAMESPACE")
+	if ns != podNamespace {
+		return "", fmt.Errorf("the peer's namespace: %s is not equal to discovery namespace: %s", ns, podNamespace)
```

Comment:
It is better to print the `advertisePeerUrl` information in the error log, especially when there may be multiple clusters under a namespace.